### ISSUE TITLE
How to debug duplicate items in queue

### ISSUE THREAD

**Issue by kaspar4** [progfolio/elpaca#167]
### Elpaca Version

(elpaca . "0afd824 HEAD -> master, origin/master, origin/HEAD")
(isntaller . 0.5)
(emacs . "GNU Emacs 29.0.92 (build 1, x86_64-pc-linux-gnu, G...")
(git . "git version 2.34.1")

### Operating System

Debian 11 (linux mint, a debian based distro)

### Description

Function `elpaca--queue` there's code that does:
```
(defun elpaca--queue (order &optional queue)
  "ADD ORDER to QUEUE or current queue. Return E."
  (let ((item (elpaca--first order)))
    (if (and (not after-init-time) (elpaca-get item))
        (warn "Duplicate item queued: %S" item)
```

I'm getting this warning for packages that I only import (through `use-package`)  only once and some of them are unlikely to be imported by others (I get this warning for magit, org-roam and ace-window).
Is there a way to debug this condition?

**Comment by progfolio** [comment]
Were you able to get this sorted, or did you still need help?

**Comment by kaspar4** [comment]
Not really root caused, but I added printing the result of `elpaca-get` for
the three packages that are giving me the warning after each of the
invocations of my own use-package wrapper.
magit and ace-window were queued during use-package of nordic-night-theme
(now I have my own theme switching code so there might be something there;
or that theme does something weird)
org-roam is queued as part of loading org-modern, so there might be
something there too.
I'll keep investigating, I was just wondering if there was a better way for
debugging these things. I closed the issue because I couldn't remove the
'bug' label and this was not a bug.

The import of the nordic theme expands to:
```
(elpaca
    (nordic-night :host sourcehut :repo "ashton314/nordic-night")
  (defvar use-package--warning108
    (function
     (lambda
       (keyword err)
       (let
           ((msg
             (format "%s/%s: %s" 'nordic-night-theme keyword
                     (error-message-string err))))
         (display-warning 'use-package msg :error)))))
  (condition-case-unless-debug err
      (if
          (not
           (require 'nordic-night-theme nil t))
          (display-warning 'use-package
                           (format "Cannot load %s" 'nordic-night-theme)
                           :error))
    (error
     (funcall use-package--warning108 :catch err))))```
```

and I don't see anything strange there, so I wonder if there's something
asynchronous going on and actually being a red herring

**Comment by progfolio** [comment]
>I'll keep investigating, I was just wondering if there was a better way for debugging these things.

I've added a better warning for the case you're running into which will tell you which package
queued the duplicate item first. See the following test case output:

<details open><summary>Test Case</summary>

```emacs-lisp
(elpaca-test
  :ref local
  :early-init (setq load-prefer-newer t)
  :init (elpaca f)
  (elpaca-wait)
  (elpaca dash))
```

</details>
<details><summary>Host Env</summary>

<table>
<tr><td>elpaca</td><td>0eadf2f HEAD -> master, origin/master, origin/HEAD</td>
<tr><td>isntaller</td><td>0.5</td>
<tr><td>emacs</td><td>GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8) of 2023-06-15</td>
<tr><td>git</td><td>git version 2.41.0</td>
</table>
</details>

<details><summary>Output</summary>

```emacs-lisp
Warning (emacs): dash previously queued as dependency of: (f)

 Test Env

Elpaca 0eadf2f HEAD -> master, origin/master, origin/HEAD
installer:      0.5
emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8)
 of 2023-06-15
git --version:  git version 2.41.0
```
</details>

I've also added an explanation for the warning to the wiki here:

https://github.com/progfolio/elpaca/wiki/Warnings#duplicate-item-warnings

Let me know if that explanation makes sense to you and the suggested advice works.

>I closed the issue because I couldn't remove the 'bug' label and this was not a bug.

Sorry for the confusion.
I've added a separate issue template for support requests.

**Comment by kaspar4** [comment]
Thanks, the new warning helps a lot. I solved one by moving, I'll take a
look at the rest when I get home. I still have:
```
⛔ Warning (emacs): magit previously queued as dependency of:
(treemacs-magit)
⛔ Warning (emacs): org-roam previously queued as dependency of:
(citar-org-roam)
```

Moving is not entirely satisfying though, as my init file is organized in
sections and moving means moving an entire section (and I'm not even sure
is a general solution as there might be conflicting requirements for
different packages).
I'm curious though, as I'm probably not understanding how elpaca works. In
all cases I had the correct `:after` constraints and I would have presumed
that this would enforce the order elpaca uses for processing things.
Again, thank you for the extended warnings.

**Comment by progfolio** [comment]
> Thanks, the new warning helps a lot.

Glad to hear it. Thank you for testing.

> Moving is not entirely satisfying though, as my init file is organized in
> sections and moving means moving an entire section (and I'm not even sure is a
> general solution as there might be conflicting requirements for different
> packages).

If the organization of your declarations is paramount, you may want to look into tangling your init.el from an Org file using the noweb syntax. This would allow you to organize things however you want, and the resulting source file can be in an order that the elisp interpreter can make use of.

Another solution, which shouldn't require too much reorganization is to minimize your usage of elpaca-wait.
That way the queue system can manage the dependencies for you. 

Assuming this is your init file: https://github.com/kaspar4/neo/blob/9c4913ef5b44a412452e684e37870cc235bfbed4/init.el
You have four calls to elpaca-wait (which creates four separate queues):

1. Sets up elpaca-use-package-mode. Makes sense since it's used throughout.
2. Sets up delight so the `:delight` use-package keyword can be used at the top-level.
3. You can remove the third call by moving the no-littering use-package declaration between 1 and 2.
4. I'm not sure why this call is necessary. What was the intent for blocking after org-modern?

I was able to successfully install everything without any of the warnings mentioned by moving no-littering and removing the third and fourth `elpaca-wait` calls.

>I'm curious though, as I'm probably not understanding how elpaca
> works. In all cases I had the correct `:after` constraints and I would have
> presumed that this would enforce the order elpaca uses for processing things.

`:after` is a use-package keyword which defers the loading of a package.
It's not used by Elpaca when installing packages.
The arguments it takes are feature names which don't necessarily align with the item names used when installing packages, so I don't think it would be trivial to implement something like that. I'd prefer to keep Elpaca's design simpler in this regard, too.

Side note:

``` emacs-lisp
;; TODO this should be possible from the corfu use-package
(elpaca
 nil
 (progn
   (add-to-list
    'load-path
    (expand-file-name "elpaca/builds/corfu/extensions"
                      user-emacs-directory))
   (require 'corfu-popupinfo)
   (setq corfu-popupinfo-delay '(1.0 . 0.3))

   (corfu-popupinfo-mode)
   (require 'corfu-history)
   (corfu-history-mode)))
```

That shouldn't be necessary if the corfu recipe is using `:files (:defaults "extensions/*")`.
e.g.


<details open><summary>Test Case</summary>

```emacs-lisp
(elpaca-test
  :init (elpaca (corfu :files (:defaults "extensions/*"))
          (global-corfu-mode)
          (corfu-popupinfo-mode)
          (message "Popup: %S" (featurep 'corfu-popupinfo)))
  ;;Blocking for non-interactive test here.
  (elpaca-wait))
```

</details>
<details><summary>Host Env</summary>

<table>
<tr><td>elpaca</td><td>615e590 HEAD -> master, origin/master, origin/HEAD</td>
<tr><td>isntaller</td><td>0.5</td>
<tr><td>emacs</td><td>GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8) of 2023-07-25</td>
<tr><td>git</td><td>git version 2.41.0</td>
</table>
</details>

<details><summary>Output</summary>

```emacs-lisp
  INFO     Scraping files for loaddefs... 
  INFO     Scraping files for loaddefs...done
  GEN      ../elpaca-autoloads.el
Cloning into '/tmp/elpaca.W3oFTR/elpaca/repos/elpaca'...
Your branch is up to date with 'origin/master'.
Checking /tmp/elpaca.W3oFTR/elpaca/repos/elpaca...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-info.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-log.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-manager.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-menu-elpa.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-menu-melpa.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-menu-org.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-process.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-test.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca-ui.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/elpaca.el...
Checking /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/doc...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/doc/early-init.el...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/doc/init.el...
Checking /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/extensions...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/extensions/elpaca-use-package.el...
Checking /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/images...
Checking /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/test...
Compiling /tmp/elpaca.W3oFTR/elpaca/repos/elpaca/test/elpaca-test.el...
Done (Total of 11 files compiled, 3 skipped in 4 directories)

Downloading MELPA recipes...
Downloading MELPA recipes...100%
Downloading NonGNU-devel ELPA...
Downloading GNU-devel ELPA...
Downloading NonGNU ELPA...
Downloading GNU ELPA...
Popup: t

 Test Env

Elpaca 615e590 HEAD -> master, origin/master, origin/HEAD
installer:      0.5
emacs-version:  GNU Emacs 30.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.38, cairo version 1.17.8)
 of 2023-07-25
git --version:  git version 2.41.0
```
</details>

**Comment by kaspar4** [comment]
Wow! Thanks a lot for looking into my init file. I'm actually moving from
.org to single init (my previous .org-mode based version is in
https://github.com/kaspar4/neo_orgmode) as I do pretty much every time I
refresh my configuration substantially). I don't see how using org would
help with ordering though, without major work.
I knew I had to take another look at the `elpaca-wait` calls. I'm new to
elpaca and when I was even greener elpaca-wait was an easy way to get out
of troubles. Ideally I should only need the one just after package
management and all additional package keyword setting modes are loaded.

If the organization of your declarations is paramount

 Is not really paramount, mostly for when I share my config with co-workers
so that they can get the parts they need. But still it would nice to be
able to declare an order, `:after` style so that the physical order can be
what makes more sense.
But I understand the desire of keeping things simple.

**Comment by progfolio** [comment]
> Wow! Thanks a lot for looking into my init file.

You're welcome.

> I don't see how using org would help with ordering though, without major work.

The pattern would generally be:


``` org-mode
* section A
:PROPERTIES:
:header-args: :noweb-ref a
:END:

#+begin_src emacs-lisp
(elpaca a)
#+end_src

* section B
:PROPERTIES:
:header-args: :noweb-ref b
:END:

#+begin_src emacs-lisp
(elpaca b)
#+end_src

* init.el
#+begin_src emacs-lisp :tangle init.el
<<b>> 
<<a>>
#+end_src
```
Note how A is presented before B in the org file, but the elisp file is tangled so B comes before A.
Of course it would mean marking up the file with Org syntax and tangling to get the resulting init file.

> it would nice to be able to declare an order, `:after` style so that the physical order can be what makes
> more sense. But I understand the desire of keeping things simple.

I'll keep it in mind, but I've yet to think of a simple way to make something like that work.

**Comment by kaspar4** [comment]
>
> Of course it would mean marking up the file with Org syntax and tangling
> to get the resulting init file.

 I see. In my org version I took a different approach and implemented an
`include` capability, so I could rather easily reorder things at that
level, but I'd still have problems if order was an issue inside each .org
file, which your approach would solve.

For now I'll stay with the single init file, as I find editing lisp in
source block annoying and it is nice to have bug-hunter bisect the file
when there're problems. Once upon a time I had nice code for editing
configs by automatically jumping to each package and narrowing to region
there. I lost that code (was pre-use package anyhow, so was file based),
but I'll rewrite it with some additions (that's the reason for the
structured section comments, they'll serve for autocompleting and narrow to
regions larger than a single package). I'll also make use of elpaca try to
install new packages and pre-populate automatically  new packages w/ all
customization variables I can  find and their default value. This will then
be called from the elpaca manager.

Thanks again for all the help.
